contracts/FeralfileVault.sol

```diff
@@ -3,34 +3,51 @@ pragma solidity ^0.8.0;
 
 import "@openzeppelin/contracts/access/Ownable.sol";
 
-contract FeralfileVault is Ownable {
-    mapping(address => bool) public exhibitionContract;
-
-    modifier onlyExhibitionContract() {
+import "./FeralfileSaleStruct.sol";
+import "./ECDSASign.sol";
+
+contract FeralfileVault is Ownable, FeralfileSaleStruct, ECDSASign {
+    mapping(bytes32 => bool) private _paidSale;
+
+    constructor(address signer_) ECDSASign(signer_) {}
+
+    /// @notice pay for buyArtwork to a FFV4 contract destination.
+    /// @param from - a FFV4 exhibition contract address
+    /// @param r_ - part of signature for validating parameters integrity
+    /// @param s_ - part of signature for validating parameters integrity
+    /// @param v_ - part of signature for validating parameters integrity
+    /// @param saleData_ - the sale data
+    function payForSale(
+        address from,
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        SaleData calldata saleData_
+    ) external {
         require(
-            exhibitionContract[msg.sender],
-            "Only exhibition contract can call this function."
+            saleData_.payByVaultContract,
+            "FeralfileVault: not pay by vault"
         );
-        _;
-    }
-
-    function setExhibitionContract(address ec) external onlyOwner {
-        exhibitionContract[ec] = true;
-    }
-
-    function unsetExhibitionContract(address ec) external onlyOwner {
-        exhibitionContract[ec] = false;
-    }
-
-    function pay(uint256 amount) external onlyExhibitionContract {
-        require(address(this).balance >= amount, "insufficient balance");
-        payable(msg.sender).transfer(amount);
+        isValidSaleData(saleData_);
```

Comment:
require(msg.sender == form) to prevent other people to run this.

Reply:
I remove the from and take msg.sender as param to make it straight